File: schemas/social-media/drizzle/post_hashtags.ts
// post_hashtags: Many-to-many link between posts and hashtags.
// See README.md for full design rationale.
import { pgTable, uuid, timestamp, index, unique } from "drizzle-orm/pg-core";
import { posts } from "./posts";
import { hashtags } from "./hashtags";

export const postHashtags = pgTable(
  "post_hashtags",
  {
    id: uuid("id").primaryKey().defaultRandom(),
    postId: uuid("post_id")
      .notNull()
      .references(() => posts.id, { onDelete: "cascade" }),
    hashtagId: uuid("hashtag_id")
      .notNull()
      .references(() => hashtags.id, { onDelete: "cascade" }),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    unique("uq_post_hashtags_post_hashtag").on(table.postId, table.hashtagId),
    index("idx_post_hashtags_hashtag_id_created_at").on(table.hashtagId, table.createdAt),
  ]
);

File: schemas/social-media/spacetimedb/post_hashtags.rs
// post_hashtags: many-to-many relationship between posts and hashtags.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

#[spacetimedb::table(name = post_hashtags, public)]
pub struct PostHashtag {
    #[primary_key]
    pub id: String, // UUID
    // Composite unique: (post_id, hashtag_id) — enforce in reducer logic.
    #[index(btree)]
    pub post_id: String, // UUID — FK → posts.id (cascade delete)
    // Composite index: (hashtag_id, created_at) — not supported in SpacetimeDB; document only.
    #[index(btree)]
    pub hashtag_id: String, // UUID — FK → hashtags.id (cascade delete)
    pub created_at: Timestamp,
}
